# Yahoo Finance Market EDA

> click [here](https://docs.google.com/document/d/1nlKlCDcs9aQovlmxv4zgyyy0BM8EfaPjDV_s8lNzu1g/edit?usp=sharing) for questions

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [2]:
df = pd.read_excel('yahoo_data.xlsx', engine='openpyxl')
df.head()

Date      Open      High       Low    Close*  Adj Close**  \
0  Apr 28, 2023  33797.43  34104.56  33728.40  34098.16     34098.16   
1  Apr 27, 2023  33381.66  33859.75  33374.65  33826.16     33826.16   
2  Apr 26, 2023  33596.34  33645.83  33235.85  33301.87     33301.87   
3  Apr 25, 2023  33828.34  33875.49  33525.39  33530.83     33530.83   
4  Apr 24, 2023  33805.04  33891.15  33726.09  33875.40     33875.40   

      Volume  
0  354310000  
1  343240000  
2  321170000  
3  297880000  
4  252020000

In [6]:
df.rename(columns = {'Open':'open', 'High':'high', 'Low':'low', 'Close*':'close', 'Adj Close**':'adj_close', 'Volume':'volume'}, inplace = True)

In [7]:
df.head()

Date      open      high       low     close  adj_close     volume
0  Apr 28, 2023  33797.43  34104.56  33728.40  34098.16   34098.16  354310000
1  Apr 27, 2023  33381.66  33859.75  33374.65  33826.16   33826.16  343240000
2  Apr 26, 2023  33596.34  33645.83  33235.85  33301.87   33301.87  321170000
3  Apr 25, 2023  33828.34  33875.49  33525.39  33530.83   33530.83  297880000
4  Apr 24, 2023  33805.04  33891.15  33726.09  33875.40   33875.40  252020000

In [9]:
# date to datetime 
df['Date'] = pd.to_datetime(df['Date'])

In [10]:
df.head()

Date      open      high       low     close  adj_close     volume
0 2023-04-28  33797.43  34104.56  33728.40  34098.16   34098.16  354310000
1 2023-04-27  33381.66  33859.75  33374.65  33826.16   33826.16  343240000
2 2023-04-26  33596.34  33645.83  33235.85  33301.87   33301.87  321170000
3 2023-04-25  33828.34  33875.49  33525.39  33530.83   33530.83  297880000
4 2023-04-24  33805.04  33891.15  33726.09  33875.40   33875.40  252020000

### How many rows and columns are present in the dataset?

In [12]:
df.shape

(1258, 7)

> 1258 rows, 7 columns

### What are the names of all columns? And what do they represent

In [13]:
df.columns

Index(['Date', 'open', 'high', 'low', 'close', 'adj_close', 'volume'], dtype='object')

> Representation
- Date: current day
- open: price of stock when stock market opened
- close: price of stock when stock market closed
- high: highest price of stock during the day
- low: least price of stock during the day
- adj_close: Adjusted closing price of the stock, after corporate actions(dividend, splits)
- volume: number of shares traded on that day

### Are there any missing values in any column?

In [14]:
df.isna().sum()

Date         0
open         0
high         0
low          0
close        0
adj_close    0
volume       0
dtype: int64

> No missing values, overall

### What is the data type of each column?

In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1258 entries, 0 to 1257
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   Date       1258 non-null   datetime64[ns]
 1   open       1258 non-null   float64       
 2   high       1258 non-null   float64       
 3   low        1258 non-null   float64       
 4   close      1258 non-null   float64       
 5   adj_close  1258 non-null   float64       
 6   volume     1258 non-null   int64         
dtypes: datetime64[ns](1), float64(5), int64(1)
memory usage: 68.9 KB


- datetime for date
- float for prices
- int for volume

### Does the dataset contain any duplicate rows?

In [19]:
df.duplicated().sum()

np.int64(0)

> no duplicates

### What is the time span of the data (earliest to latest trading day)?

In [21]:
print(df['Date'].dt.year.min())
print(df['Date'].dt.year.max())

2018
2023


> from 2018 to 2023

### Is the date stored as a column or as an index?

In [22]:
df.columns

Index(['Date', 'open', 'high', 'low', 'close', 'adj_close', 'volume'], dtype='object')

> as a column

### Are weekends or non-trading days present in the dataset?

In [31]:
df['Date'].dt.day_name() == 'Saturday'

0       False
1       False
2       False
3       False
4       False
        ...  
1253    False
1254    False
1255    False
1256    False
1257    False
Name: Date, Length: 1258, dtype: bool

In [32]:
df['Date'].dt.day_name() == 'Sunday'

0       False
1       False
2       False
3       False
4       False
        ...  
1253    False
1254    False
1255    False
1256    False
1257    False
Name: Date, Length: 1258, dtype: bool

> No, Weekend/non-trading day is present 

### What is the basic statistical summary (mean, min, max, std) of numerical columns?

In [34]:
df.describe().T

count                           mean                  min  \
Date         1258  2020-10-28 09:39:12.305246464  2018-05-01 00:00:00   
open       1258.0                   29595.823045             19028.36   
high       1258.0                   29776.945739             19121.01   
low        1258.0                   29402.432226             18213.65   
close      1258.0                   29599.361677             18591.93   
adj_close  1258.0                   29599.361677             18591.93   
volume     1258.0               345063624.801272           86150000.0   

                           25%                  50%                  75%  \
Date       2019-07-31 06:00:00  2020-10-27 12:00:00  2022-01-26 18:00:00   
open                26041.2675             29201.41           33604.0275   
high                 26163.155            29335.685            33825.445   
low                 25877.8725              28996.5           33346.8275   
close                 26027.12             29199.46           33600.3425   
adj_close             26027.12             29199.46           33600.3425   
volume             277312500.0          324725000.0          387510000.0   

                           max               std  
Date       2023-04-28 00:00:00               NaN  
open                   36722.6       4006.078299  
high                  36952.65       4009.007573  
low                    36636.0       4004.949066  
close                 36799.65       4007.468822  
adj_close             36799.65       4007.468822  
volume             915990000.0  106914195.136917